File: examples/SDK/python/data_plane/WorkingWithQueries/query-return-consumed-capacity.py
from __future__ import print_function  # Python 2/3 compatibility
import boto3
import json
import decimal
from boto3.dynamodb.conditions import Key, Attr
from botocore.exceptions import ClientError


# Helper class to convert a DynamoDB item to JSON.
class DecimalEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, decimal.Decimal):
            return str(o)
        return super(DecimalEncoder, self).default(o)


dynamodb = boto3.resource('dynamodb', region_name='us-west-2')

table = dynamodb.Table('Music')

try:
    response = table.query(
        ExpressionAttributeNames={
            '#pk': 'Artist',
        },
        ExpressionAttributeValues={
            ':pk': 'Michael Jackson',
        },
        KeyConditionExpression='#pk = :pk',
        ReturnConsumedCapacity='TOTAL'
    )
except ClientError as e:
    print(e.response['Error']['Message'])
else:
    print('Consumed capacity for for this query:', json.dumps(
        response['ConsumedCapacity'], indent=4, cls=DecimalEncoder))

    if 'LastEvaluatedKey' in response:
        print('Not all items have been retrieved by this query. At least one another request is required to get all available items. The last evaluated key corresponds to',
              json.dumps(response['LastEvaluatedKey'], indent=4, cls=DecimalEncoder))

    print('Query response:', json.dumps(response, indent=4, cls=DecimalEncoder))

File: examples/SDK/node.js/sdk_v3/data_plane/WorkingWithQueries/query-return-consumed-capacity.js
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand } from "@aws-sdk/lib-dynamodb";

const client = new DynamoDBClient({ region: "us-west-2" });
const docClient = DynamoDBDocumentClient.from(client);

const query = async () => {
  const params = {
    TableName: "Music",
    KeyConditionExpression: "#pk = :pk",
    ExpressionAttributeNames: {
      "#pk": "Artist",
    },
    ExpressionAttributeValues: {
      ":pk": "Michael Jackson",
    },
    ReturnConsumedCapacity: "TOTAL",
  };

  try {
    // Send the query command
    const response = await docClient.send(new QueryCommand(params));

    // Check if there are more results to retrieve
    if (response.LastEvaluatedKey) {
      console.log(
        `Not all items have been retrieved by this query. At least one another request is required to get all available items. The last evaluated key corresponds to ${JSON.stringify(response.LastEvaluatedKey)}.`
      );
    }

    console.log(`Consumed capacity for for this query: ${JSON.stringify(response.ConsumedCapacity)}`);

    return response;
  } catch (error) {
    console.error("Error querying DynamoDB:", error);
    throw error;
  }
};

query()
  .then((response) => console.log(`Query response: ${JSON.stringify(response, null, 2)}`))
  .catch((error) => console.error(JSON.stringify(error, null, 2)));
